### PR TITLE
`<type_traits>`: Workaround for `common_reference`

### LINKED ISSUE #2581: `<iterator>`: `std::contiguous_iterator<const volatile int*> == false` because of `volatile`?
`static_assert(std::contiguous_iterator<const volatile int*>)` fails with this STL, while it works with lib(std)c++. Without the `volatile` it works everywhere. Who is correct?
I can't see why a pointer to `volatile` data shouldn't be a valid iterator.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #2581

Adds workaround for these cases:
When I is a cv-unqualified scalar type, `LRef` is _`cv`_`I&`, `RRef` is _`cv`_`I&&`, MSVC generally erroneously computes `_Cond_res<LRef, RRef>` as `I&`, while the right result is `I`. Although sometimes pointer types get right results, and it seems that `I&&` doesn't trigger the bug.

`_Cond_res<volatile int&, const int&>` is not fixed yet, but `std::common_reference_t<volatile int&, const int&>` doesn't rely on its result.

Additional workarounds:
- MSVC also sometimes computes `_Cond_res<A&, A&&>` incorrectly to `A&` when `A` is an array type. The result should be decayed.
- MSVC computes the composite pointer of `volatile T(*)[N]` and `const T(*)[N]` to ill-defined result type `T volatile (const volatile*)[N]`. Currently my workaround is incomplete. A complete workaround may need to hack the whole computation in [[conv.qual]/3](https://eel.is/c++draft/conv.qual#3).
- DevCom-1627396 matters.

### CODE REVIEW

File: tests/std/tests/GH_002581_common_reference_workaround/test.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include <cstddef>
```
Comment (fsb4000) [inline tests/std/tests/GH_002581_common_reference_workaround/test.cpp:1]: you forgot the comment:
```c++
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

Also we should modify the file:
https://github.com/microsoft/STL/blob/ff5445078a73b5e9445dffc656028a47fa790509/tests/std/test.lst

Comment (frederick-vs-ja) [reply tests/std/tests/GH_002581_common_reference_workaround/test.cpp:1]: Thanks! I'll add to these files when sufficient workarounds are done.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I went ahead and pushed changes for the really minor things I noticed.

Generally, I am allergic to *permanently* working around compiler deficiencies in type traits (as our ancient implementation of `is_pod` attempted to do) - that's bad for throughput, it's hard to know if the workaround is truly complete, and it accretes complexity as the language evolves. However, a compiler-specific, well-commented, well-tested, temporary workaround to a fundamental type trait that has been blocking a number of libcxx tests from running, seems like an entirely reasonable thing to do. (It's consistent with our usual practice of working around compiler bugs while reporting them and waiting for fixes.)

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` to resolve a trivial adjacent-add conflict in `test.lst`.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]
I'm mildly concerned that we'll discover some corner where trait(s) that are intended to reflect language behavior disagree with what the compiler actually does. Defining `is_convertible_v<int, int(*)(float)>` to `true`, for example, doesn't magically make that conversion work. "Lying" about reflection doesn't really fix a bug so much as relocate it. I can find no such issues here with a bit of cursory investigation, however, so I think it's safe to merge this.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving this type traits machinery to be more correct than the compiler! :joy_cat: :rocket: :tada:

**Pull request merged**